ibm/resource_ibm_pi_image.go

```diff
@@ -63,145 +122,193 @@ func resourceIBMPIImage() *schema.Resource {
-	if err != nil {
-		return false, err
-	}
-	parts, err := idParts(d.Id())
-	if err != nil {
-		return false, err
-	}
-	if len(parts) < 2 {
-		return false, fmt.Errorf("Incorrect ID %s: Id should be a combination of powerInstanceID/ImageID", d.Id())
-	}
-	name := parts[1]
-	powerinstanceid := parts[0]
-	client := st.NewIBMPIImageClient(sess, powerinstanceid)
-
-	image, err := client.Get(parts[1], powerinstanceid)
-	if err != nil {
-		if apiErr, ok := err.(bmxerror.RequestFailure); ok {
-			if apiErr.StatusCode() == 404 {
-				return false, nil
-			}
-		}
-		return false, fmt.Errorf("Error communicating with the API: %s", err)
-	}
-	return *image.ImageID == name, nil
-}
-
-func isWaitForIBMPIImageAvailable(client *st.IBMPIImageClient, id string, timeout time.Duration, powerinstanceid string) (interface{}, error) {
+func isWaitForIBMPIImageAvailable(ctx context.Context, client *st.IBMPIImageClient, id string, timeout time.Duration, powerinstanceid string) (interface{}, error) {
 	log.Printf("Waiting for Power Image (%s) to be available.", id)
 
 	stateConf := &resource.StateChangeConf{
 		Pending:    []string{"retry", helpers.PIImageQueStatus},
 		Target:     []string{helpers.PIImageActiveStatus},
-		Refresh:    isIBMPIImageRefreshFunc(client, id, powerinstanceid),
+		Refresh:    isIBMPIImageRefreshFunc(ctx, client, id, powerinstanceid),
 		Timeout:    timeout,
 		Delay:      20 * time.Second,
 		MinTimeout: 10 * time.Second,
 	}
 
-	return stateConf.WaitForState()
+	return stateConf.WaitForStateContext(ctx)
 }
 
-func isIBMPIImageRefreshFunc(client *st.IBMPIImageClient, id, powerinstanceid string) resource.StateRefreshFunc {
+func isIBMPIImageRefreshFunc(ctx context.Context, client *st.IBMPIImageClient, id, powerinstanceid string) resource.StateRefreshFunc {
 
 	log.Printf("Calling the isIBMPIImageRefreshFunc Refresh Function....")
 	return func() (interface{}, string, error) {
-		image, err := client.Get(id, powerinstanceid)
+		image, err := client.GetWithContext(ctx, id, powerinstanceid)
 		if err != nil {
 			return nil, "", err
 		}
 
 		if image.State == "active" {
-
 			return image, helpers.PIImageActiveStatus, nil
 		}
 
 		return image, helpers.PIImageQueStatus, nil
 	}
 }
+
+func waitForIBMPIJobCompleted(ctx context.Context, client *st.IBMPIJobClient, jobID, cloudInstanceID string, timeout time.Duration) (interface{}, error) {
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{helpers.JobStatusQueued, helpers.JobStatusReadyForProcessing, helpers.JobStatusInProgress, helpers.JobStatusRunning, helpers.JobStatusWaiting},
+		Target:  []string{helpers.JobStatusCompleted, helpers.JobStatusFailed},
+		Refresh: func() (interface{}, string, error) {
+			job, err := client.GetWithContext(ctx, jobID, cloudInstanceID)
+			if err != nil {
+				log.Printf("[DEBUG] get job failed %v", err)
+				return nil, "", fmt.Errorf(errors.GetJobOperationFailed, jobID, err)
+			}
+			if *job.Status.State == helpers.JobStatusFailed {
```

Comment:
Please add check for job != nil && job.Status != nil && job.Status.State == helpers.JobStatusFailed just to eliminate any nil pointer crash